gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java

```diff
@@ -0,0 +1,231 @@
+      // For the system configuration they should be kept, like ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY
+      if (rawProp.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_PATH_KEY)) {
+        rawProp.remove(ConfigurationKeys.JOB_CONFIG_FILE_PATH_KEY);
+      }
+      if (rawProp.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY)) {
+        rawProp.remove(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY);
+      }
+    }
+    optionalProp = Optional.fromNullable(rawProp);
+    config = Optional.fromNullable(ConfigUtils.propertiesToConfig(rawProp));
+
+    // The builder has null-checker. Leave the checking there.
+    return JobSpec.builder(jobConfigURI)
+        .withConfigAsProperties(optionalProp.get())
+        .withConfig(config.get())
+        .withDescription(description.get())
+        .withVersion(version)
+        .build();
+  }
+
+  public static Set<String> getJobConfigurationFileExtensions(Properties properties) {
+    Iterable<String> jobConfigFileExtensionsIterable = Splitter.on(",")
+        .omitEmptyStrings()
+        .trimResults()
+        .split(properties.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_EXTENSIONS_KEY,
+            ConfigurationKeys.DEFAULT_JOB_CONFIG_FILE_EXTENSIONS));
+    return ImmutableSet.copyOf(Iterables.transform(jobConfigFileExtensionsIterable, new Function<String, String>() {
+      @Override
+      public String apply(String input) {
+        return null != input ? input.toLowerCase() : "";
+      }
+    }));
+  }
+
+  /**
+   * Create and attach {@link PathAlterationDetector}s for the given
+   * root directory and any nested subdirectories under the root directory to the given
+   * {@link PathAlterationDetector}.
+   * @param detector  a {@link PathAlterationDetector}
+   * @param listener a {@link gobblin.util.filesystem.PathAlterationListener}
+   * @param observerOptional Optional observer object. For testing routine, this has been initialized by user.
+   *                         But for general usage, the observer object is created inside this method.
+   * @param rootDirPath root directory
+   * @throws IOException
+   */
+  public static void addPathAlterationObserver(PathAlterationDetector detector, PathAlterationListener listener,
+      Optional<PathAlterationObserver> observerOptional, Path rootDirPath)
+      throws IOException {
+    PathAlterationObserver observer;
+    if (observerOptional.isPresent()) {
+      observer = observerOptional.get();
+    } else {
+      observer = new PathAlterationObserver(rootDirPath);
+    }
+    observer.addListener(listener);
+    detector.addObserver(observer);
+  }
+
+  /**
+   * @param jobConfigDirPath The root directory of job configuration files.
+   * @param path Unknown for whether it is full path or not. Will need to check this first.
+   * @return Resovled full path.
+   */
+  public static Path resovledRelativePath(Path jobConfigDirPath, Path path) {
```

Comment:
why do you need this? There should be a clear division: user generated paths are always "relative", while internal paths always start with `jobConfigDirPath`.